### PR TITLE
FIX #5382 merge error with same category on societe object

### LINKED ISSUE #5382: Merge of third parties does not work when both are in the same category
When two companies have the same category assigned, the merge of the companies fails with a DB_ERROR_RECORD_ALREADY_EXISTS error.
## Environment
- **Version**: 3.9.1
- **OS**: Debian Jessie 8
- **Web server**: Apache 2.4
- **PHP**: 5.4.45
- **Database**: MySQL 5.5
- **URL(s)**: /societe/soc.php?action=merge&socid=5820
## Expected and actual behavior

Merging of companies should work, independent of assigned categories.
Currently when both companies have the same category assigned, it fails to merge the companies.
## Steps to reproduce the behavior
1. Create two companies
2. Create a Category
3. Assign the same category to both companies
4. Now merge both companies
## LOG

In this example I try to merge company 5820 with 5828
Both companys have the Category 14 assigned, which then leads to the transaction failure

2016-06-21 17:38:20 DEBUG   2001:470:26:421:41cc:bbca:c398:f566 BEGIN Transaction
2016-06-21 17:38:20 DEBUG   2001:470:26:421:41cc:bbca:c398:f566  sql=UPDATE llx_adherent SET fk_soc = 5820 WHERE fk_soc = 5828
2016-06-21 17:38:20 DEBUG   2001:470:26:421:41cc:bbca:c398:f566  sql=SELECT rowid
FROM llx_societe_commerciaux
WHERE fk_soc = 5820 AND fk_user IN (
  SELECT fk_user
  FROM llx_societe_commerciaux
  WHERE fk_soc = 5828
);
2016-06-21 17:38:20 DEBUG   2001:470:26:421:41cc:bbca:c398:f566  sql=UPDATE llx_societe_address SET fk_soc = 5820 WHERE fk_soc = 5828
2016-06-21 17:38:20 DEBUG   2001:470:26:421:41cc:bbca:c398:f566  sql=UPDATE llx_societe_commerciaux SET fk_soc = 5820 WHERE fk_soc = 5828
2016-06-21 17:38:20 DEBUG   2001:470:26:421:41cc:bbca:c398:f566  sql=UPDATE llx_societe_log SET fk_soc = 5820 WHERE fk_soc = 5828
2016-06-21 17:38:20 DEBUG   2001:470:26:421:41cc:bbca:c398:f566  sql=UPDATE llx_societe_prices SET fk_soc = 5820 WHERE fk_soc = 5828
2016-06-21 17:38:20 DEBUG   2001:470:26:421:41cc:bbca:c398:f566  sql=UPDATE llx_societe_remise SET fk_soc = 5820 WHERE fk_soc = 5828
2016-06-21 17:38:20 DEBUG   2001:470:26:421:41cc:bbca:c398:f566  sql=UPDATE llx_societe_remise_except SET fk_soc = 5820 WHERE fk_soc = 5828
2016-06-21 17:38:20 DEBUG   2001:470:26:421:41cc:bbca:c398:f566  sql=UPDATE llx_societe_rib SET fk_soc = 5820 WHERE fk_soc = 5828
2016-06-21 17:38:20 DEBUG   2001:470:26:421:41cc:bbca:c398:f566  sql=UPDATE llx_bookmark SET fk_soc = 5820 WHERE fk_soc = 5828
2016-06-21 17:38:20 DEBUG   2001:470:26:421:41cc:bbca:c398:f566  sql=UPDATE llx_categorie_societe SET fk_soc = 5820 WHERE fk_soc = 5828
2016-06-21 17:38:20 ERR     2001:470:26:421:41cc:bbca:c398:f566  DoliDBMysqli::query SQL Error message: DB_ERROR_RECORD_ALREADY_EXISTS Duplicate entry '14-5820' for key 'PRIMARY'
2016-06-21 17:38:20 DEBUG   2001:470:26:421:41cc:bbca:c398:f566 ROLLBACK Transaction
## Suggested steps

On merge first check if the categories are already assigned to the target company

### PR DESCRIPTION
**Pull request by ghost**
- avoid merge error on category on company fusion due to duplicate index. Not the best way, but the simple with actual call,
- little fix on sql string with prefix table

### CODE REVIEW

File: htdocs/societe/class/societe.class.php [hunk 1]
```diff
@@ -3483,17 +3483,17 @@ public static function replaceThirdparty(DoliDB $db, $origin_id, $dest_id)
...
-FROM llx_societe_commerciaux
-WHERE fk_soc = '.(int) $dest_id.' AND fk_user IN (
-  SELECT fk_user
-  FROM llx_societe_commerciaux
-  WHERE fk_soc = '.(int) $origin_id.'
-);';
+				FROM '.MAIN_DB_PREFIX.'societe_commerciaux
+				WHERE fk_soc = '.(int) $dest_id.' AND fk_user IN (
+				  SELECT fk_user
+				  FROM llx_societe_commerciaux
```
Comment (ghost) [inline htdocs/societe/class/societe.class.php]: Shouldn't use MAIN_DB_PREFIX  ?

Comment (ghost) [inline htdocs/societe/class/societe.class.php]: Don't see this other bad use when i correct first one ;)

File: htdocs/core/class/commonobject.class.php [hunk 2]
```diff
@@ -4181,16 +4181,25 @@ public function getRights()
...
-	public static function commonReplaceThirdparty(DoliDB $db, $origin_id, $dest_id, array $tables)
+	public static function commonReplaceThirdparty(DoliDB $db, $origin_id, $dest_id, array $tables, $delete_if_error = false)
 	{
 		foreach ($tables as $table)
 		{
 			$sql = 'UPDATE '.MAIN_DB_PREFIX.$table.' SET fk_soc = '.$dest_id.' WHERE fk_soc = '.$origin_id;
 
 			if (!$db->query($sql)) {
-				return false;
+				if($delete_if_error) {
```
Comment (eldy) [inline htdocs/core/class/commonobject.class.php:4194]: This looks not enough. If there is categ A-B on kept thirdarty and B-C on old one, we must get A-B-C after merge. Not A-B. A special test on table == llx_categories seems required to include a DELETE of category on second thirparty that is already on first one before running the update is required.

**Comment by eldy** [comment]
It seems this PR was set with "Discussion" or "PR to fix" tag, but no answer nor correction was provided since a long time. So request is automatically closed.
Please reopened and add a comment if you think this is an error.

**Pull request closed without merging**